index_test.go

```diff
@@ -2261,3 +2262,246 @@ func TestIndexMappingDocValuesDynamic(t *testing.T) {
+		t.Error(err)
+	}
+
+	docy := map[string]interface{}{
+		"name": "jasper",
+		"desc": "clojure",
+	}
+	err = idx.Index("y", docy)
+	if err != nil {
+		t.Error(err)
+	}
+
+	err = idx.Delete("y")
+	if err != nil {
+		t.Error(err)
+	}
+
+	docx := map[string]interface{}{
+		"name": "rose",
+		"desc": "xoogler",
+	}
+	err = idx.Index("x", docx)
+	if err != nil {
+		t.Error(err)
+	}
+
+	err = idx.SetInternal([]byte("status"), []byte("pending"))
+	if err != nil {
+		t.Error(err)
+	}
+
+	docb := map[string]interface{}{
+		"name": "sree",
+		"desc": "cbft janitor",
+	}
+	batch := idx.NewBatch()
+	err = batch.Index("b", docb)
+	if err != nil {
+		t.Error(err)
+	}
+	batch.Delete("x")
+	batch.SetInternal([]byte("batchi"), []byte("batchv"))
+	batch.DeleteInternal([]byte("status"))
+	err = idx.Batch(batch)
+	if err != nil {
+		t.Error(err)
+	}
+	val, err := idx.GetInternal([]byte("batchi"))
+	if err != nil {
+		t.Error(err)
+	}
+	if string(val) != "batchv" {
+		t.Errorf("expected 'batchv', got '%s'", val)
+	}
+	val, err = idx.GetInternal([]byte("status"))
+	if err != nil {
+		t.Error(err)
+	}
+	if val != nil {
+		t.Errorf("expected nil, got '%s'", val)
+	}
+
+	err = idx.SetInternal([]byte("seqno"), []byte("7"))
+	if err != nil {
+		t.Error(err)
+	}
+	err = idx.SetInternal([]byte("status"), []byte("ready"))
+	if err != nil {
+		t.Error(err)
+	}
+	err = idx.DeleteInternal([]byte("status"))
+	if err != nil {
+		t.Error(err)
+	}
+	val, err = idx.GetInternal([]byte("status"))
+	if err != nil {
+		t.Error(err)
+	}
+	if val != nil {
+		t.Errorf("expected nil, got '%s'", val)
+	}
+
+	val, err = idx.GetInternal([]byte("seqno"))
+	if err != nil {
+		t.Error(err)
+	}
+	if string(val) != "7" {
+		t.Errorf("expected '7', got '%s'", val)
+	}
+
+	count, err := idx.DocCount()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if count != 2 {
+		t.Errorf("expected doc count 2, got %d", count)
+	}
+
+	doc, err := idx.Document("a")
+	if err != nil {
+		t.Fatal(err)
+	}
+	foundNameField := false
+	doc.VisitFields(func(field index.Field) {
+		if field.Name() == "name" && string(field.Value()) == "tester" {
+			foundNameField = true
+		}
+	})
+	if !foundNameField {
+		t.Errorf("expected to find field named 'name' with value 'tester'")
+	}
+
+	fields, err := idx.Fields()
+	if err != nil {
+		t.Fatal(err)
+	}
+	expectedFields := map[string]bool{
+		"_all": false,
+		"name": false,
+		"desc": false,
+	}
+	if len(fields) < len(expectedFields) {
+		t.Fatalf("expected %d fields got %d", len(expectedFields), len(fields))
+	}
+	for _, f := range fields {
+		expectedFields[f] = true
+	}
+	for ef, efp := range expectedFields {
+		if !efp {
+			t.Errorf("field %s is missing", ef)
+		}
+	}
+
+	// now create a copy of the index, and repeat assertions on it
+	copyableIndex, ok := idx.(IndexCopyable)
+	if !ok {
+		t.Fatal("index doesn't support copy")
+	}
+	backupIndexPath := createTmpIndexPath(t)
+	defer cleanupTmpIndexPath(t, backupIndexPath)
+
+	getWriter := func(filePath string) io.WriteCloser {
```

Comment:
So, I still strongly dislike the design that requires the caller to pass this function to CopyTo().

Let me start by asking, what would a user expect the API to look like?  I feel pretty confident it would be something like:

```
func (b *bleveIndexImplStruct) CopyTo(destinationPath string) error
```

The end-user doesn't expect to pass a function that creates writers inside that directory, we should do that part on their behalf.

What if all of `getWriter()` here were internal to bleve.  And then when the user calls CopyTo(path), we internally use this getWriter implementation?  It seems like the only logic inside getWriter has to do with internal implemenetation choices made by bleve about file/dir naming.

Reply:
@mschoch, I completely agree that the current API is CB-specific and can't be used by other users due to its ugliness.

The reason I ended up here is that - 

If it accepts a path, then cbft has to copy/duplicate all the disk segment files into a temp location. And then perform a stream write.  Wanted to save the additional disk requirements as well as the extra stream write/copy. 
With the current api, we can directly write to stream without demanding additional disk resources. 

Not sure whether we can achieve that if we accept a CopyTo(destinationPath string)?

Reply:
OK, so I'm going to try to make it work, because I think that getting the API right is the most important thing.

Reply:
@sreekanth-cb why can't this work?  https://play.golang.org/p/vqwsUB7h487

Reply:
I see what you meant now. 
CBFT can directly call the scorch level implementation which takes a getWriter and bleve users can invoke the Index level API which takes a path.
Sounds good! 👍

Reply:
@sreekanth-cb I'm still not clear what cbft does differently, such that it needs to do that?  Can you link me to the right part of the cbft review?  I did try to find, but did not...

Reply:
hm,  Looks like I am circling around the same problems here. 

The above approach won't help as the index impl's meta data is associated with the bleve.Index implementation and my use case can't get that copied by directly calling any APIs from the index implementations like scorch which takes a getWriter() callback.

Then the only way would be to directly call index level CopyTo(path) which won't solve the additional disk space requirements. Let's sync on this in a call.